File: src/main/java/com/example/librarymanagement/service/impl/JournalServiceImpl.java
package com.example.librarymanagement.service.impl;

import com.example.librarymanagement.exception.EntityNotFoundException;
import com.example.librarymanagement.model.entity.Journal;
import com.example.librarymanagement.repository.JournalRepository;
import com.example.librarymanagement.service.JournalService;
import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class JournalServiceImpl implements JournalService {
    private final JournalRepository journalRepository;

    public JournalServiceImpl(JournalRepository journalRepository) {
        this.journalRepository = journalRepository;
    }

    @Override
    public Journal createJournal(Journal journal) {
        return journalRepository.save(journal);
    }

    @Override
    public Journal getJournalById(Long id) {
        return journalRepository.findById(id).orElseThrow(
                () -> new EntityNotFoundException("Journal"));
    }

    @Override
    public List<Journal> getJournalByBookPresenceId(Long bookPresenceId) {
        return journalRepository.findByBookPresenceId(bookPresenceId);
    }

    @Override
    public Journal getJournalByBookPresenceIdAndUserId(Long bookPresenceId, Long userId) {
        return journalRepository.findByBookPresenceIdAndUserId(bookPresenceId, userId).orElseThrow(
                () -> new EntityNotFoundException("Journal"));
    }

    @Override
    public List<Journal> getJournalByUserId(Long userId) {
        return journalRepository.findAllByUserId(userId);
    }

    @Override
    public void deleteJournal(Long id) {
        journalRepository.findById(id).ifPresent(journalRepository::delete);
    }
}

File: src/main/java/com/example/librarymanagement/service/impl/JournalServiceImpl.kt
package com.example.librarymanagement.service.impl

import com.example.librarymanagement.exception.EntityNotFoundException
import com.example.librarymanagement.model.domain.Journal
import com.example.librarymanagement.repository.JournalRepository
import com.example.librarymanagement.service.JournalService
import org.springframework.stereotype.Service

@Service
class JournalServiceImpl(private val journalRepository: JournalRepository) : JournalService {
    override fun createJournal(journal: Journal): Journal = journalRepository.save(journal)

    override fun getJournalById(id: String): Journal {
        return journalRepository.findById(id) ?: throw EntityNotFoundException("Journal")
    }

    override fun updateJournal(id: String, updatedJournal: Journal): Journal {
        val journal = getJournalById(id).apply { this.dateOfReturning = updatedJournal.dateOfReturning }
        return journalRepository.save(journal)
    }

    override fun findByBookPresenceIdAndUserIdAndDateOfReturningIsNull(bookPresenceId: String, userId: String): Journal {
        return journalRepository.findByBookPresenceIdAndUserIdAndDateOfReturningIsNull(bookPresenceId, userId)
            ?: throw EntityNotFoundException("Journal")
    }

    override fun getJournalByUserId(userId: String): List<Journal> = journalRepository.findAllByUserId(userId)
}
